File: packages/xyd-opensdk-java/src/javatype.ts
import type { Field, NamedType, TypeRef } from '@xyd-js/opensdk-core';

import { pascalCase } from './naming';

/** The mapped discriminator of a union (propertyName + non-empty mapping), else null. */
export function unionMapping(type: NamedType): { property: string; mapping: Record<string, string> } | null {
  if (type.kind !== 'union') return null;
  const disc = type.discriminator;
  if (!disc?.propertyName || !disc.mapping || Object.keys(disc.mapping).length === 0) return null;
  return { property: disc.propertyName, mapping: disc.mapping };
}

/** Whether a field's type is a fixed-literal scalar (JSON Schema const / 1-value enum). */
export function isConstField(field: Field): boolean {
  const ref = field.type as TypeRef | undefined;
  return ref?.kind === 'scalar' && ref.const !== undefined;
}

/** The Java literal for a const scalar value (string quoted, number/bool verbatim). */
export function constLiteral(value: unknown): string {
  if (typeof value === 'number') return `${value}L`;
  if (typeof value === 'boolean') return String(value);
  return JSON.stringify(String(value));
}

/**
 * Whether a TypeRef ultimately carries binary bytes (`format: binary`),
 * following array items and named union/alias refs (struct/enum refs are never
 * binary). Mirrors the Python emitter's isBinaryTypeRef — a json-labelled body
 * that hides a binary field must still route through the multipart encoder.
 */
export function isBinaryTypeRef(
  ref: TypeRef | undefined,
  types: Map<string, NamedType>,
  seen: Set<string> = new Set(),
): boolean {
  if (!ref) return false;
  if (ref.kind === 'scalar') return ref.scalar === 'string' && ref.format === 'binary';
  if (ref.kind === 'array') return isBinaryTypeRef(ref.items as TypeRef | undefined, types, seen);
  if (ref.kind === 'ref' && ref.name) {
    if (seen.has(ref.name)) return false;
    seen.add(ref.name);
    const named = types.get(ref.name);
    if (!named) return false;
    if (named.kind === 'union') return (named.variants || []).some((v) => isBinaryTypeRef(v as TypeRef, types, seen));
    if (named.kind === 'alias') return isBinaryTypeRef(named.of as TypeRef | undefined, types, seen);
  }
  return false;
}

/**
 * How a query param serializes, by resolved type shape (mirrors the Go emitter's
 * queryKind): scalar/enum -> single key=value; array -> repeated/comma-joined;
 * map -> deepObject or JSON; struct ref / any -> JSON-encoded.
 */
export type QueryKind = 'scalar' | 'array' | 'map' | 'object';

export function queryKind(ref: TypeRef | undefined, types: Map<string, NamedType>): QueryKind {
  if (!ref) return 'scalar';
  if (ref.kind === 'scalar') return 'scalar';
  if (ref.kind === 'array') return 'array';
  if (ref.kind === 'map') return 'map';
  if (ref.kind === 'ref' && ref.name) {
    const named = types.get(ref.name);
    if (named?.kind === 'enum') return 'scalar';
    if (named?.kind === 'alias') return queryKind(named.of as TypeRef | undefined, types);
    return 'object';
  }
  return 'object';
}

/**
 * Map an IR TypeRef to a Java type expression. Named refs resolve to their
 * generated class name (struct/enum) in the root package; aliases inline their
 * underlying type; unions/unknowns fall back to `Object` (a PHASE-2 seam).
 * Scalars map to boxed builtins so every value is nullable-friendly.
 */
export function javaType(ref: TypeRef | undefined, types: Map<string, NamedType>): string {
  if (!ref) return 'Object';
  switch (ref.kind) {
    case 'scalar':
      return javaScalar(ref.scalar, ref.format);
    case 'array':
      return `List<${javaType(ref.items as TypeRef | undefined, types)}>`;
    case 'map':
      return `Map<String, ${javaType(ref.values as TypeRef | undefined, types)}>`;
    case 'ref':
      return javaRefType(ref.name, types);
    default:
      return 'Object';
  }
}

function javaScalar(scalar: string | undefined, format: string | undefined): string {
  switch (scalar) {
    case 'string':
      return format === 'binary' ? 'byte[]' : 'String';
    case 'integer':
      return 'Long';
    case 'number':
      return 'Double';
    case 'boolean':
      return 'Boolean';
    default:
      return 'Object';
  }
}

function javaRefType(name: string | undefined, types: Map<string, NamedType>): string {
  if (!name) return 'Object';
  const named = types.get(name);
  if (!named) return 'Object'; // unknown placeholder ref -> raw JSON value
  if (named.kind === 'struct' || named.kind === 'enum') return pascalCase(name);
  if (named.kind === 'alias') return javaType(named.of as TypeRef | undefined, types);
  return 'Object'; // union
}

/** Whether a TypeRef is a binary blob (string format=binary) — an upload/download payload. */
export function isBinaryRef(ref: TypeRef | undefined): boolean {
  return ref?.kind === 'scalar' && ref.scalar === 'string' && ref.format === 'binary';
}

/** Whether a ref resolves to a generated class that carries `fromJson`/`toJsonMap` (struct/enum). */
export function isModelRef(ref: TypeRef | undefined, types: Map<string, NamedType>): boolean {
  if (ref?.kind !== 'ref' || !ref.name) return false;
  const named = types.get(ref.name);
  return named?.kind === 'struct' || named?.kind === 'enum';
}

/** The lambda variable for a decode at nesting `depth` — `__e`, `__e1`, `__e2`, ...
 * A nested array/map decode must NOT reuse its enclosing lambda's parameter
 * (Java forbids shadowing a lambda param), so each level gets a distinct name.
 * Depth 0 stays `__e` to keep the single-level (overwhelmingly common) shape. */
function lambdaVar(depth: number): string {
  return depth === 0 ? '__e' : `__e${depth}`;
}

/**
 * A Java expression that decodes a parsed-JSON `Object` (`src`) into the typed
 * value for `ref`. Reused by model `fromJson` and method-return decoding, so
 * both stay in lockstep. Arrays/maps use vendored `Json.mapList`/`mapValues`
 * with a lambda over each element; struct/enum refs call their `fromJson`.
 * `depth` scopes the lambda variable so nested containers don't shadow.
 */
export function javaDecode(ref: TypeRef | undefined, src: string, types: Map<string, NamedType>, depth = 0): string {
  if (!ref) return src;
  switch (ref.kind) {
    case 'scalar':
      return scalarDecode(ref.scalar, ref.format, src);
    case 'array': {
      const v = lambdaVar(depth);
      return `Json.mapList(${src}, ${v} -> ${javaDecode(ref.items as TypeRef | undefined, v, types, depth + 1)})`;
    }
    case 'map': {
      const v = lambdaVar(depth);
      return `Json.mapValues(${src}, ${v} -> ${javaDecode(ref.values as TypeRef | undefined, v, types, depth + 1)})`;
    }
    case 'ref':
      return refDecode(ref.name, src, types);
    default:
      return src;
  }
}

function scalarDecode(scalar: string | undefined, format: string | undefined, src: string): string {
  switch (scalar) {
    case 'string':
      return format === 'binary' ? `Json.asBytes(${src})` : `Json.asString(${src})`;
    case 'integer':
      return `Json.asLong(${src})`;
    case 'number':
      return `Json.asDouble(${src})`;
    case 'boolean':
      return `Json.asBoolean(${src})`;
    default:
      return src;
  }
}

function refDecode(name: string | undefined, src: string, types: Map<string, NamedType>): string {
  if (!name) return src;
  const named = types.get(name);
  if (!named) return src;
  if (named.kind === 'struct' || named.kind === 'enum') return `${pascalCase(name)}.fromJson(${src})`;
  if (named.kind === 'alias') return javaDecode(named.of as TypeRef | undefined, src, types);
  // A mapped discriminated union decodes to its concrete variant via its
  // generated helper class; a mapping-less union keeps the raw parsed value.
  if (named.kind === 'union' && unionMapping(named)) return `${pascalCase(name)}.fromJson(${src})`;
  return src; // open union
}

File: crates/xyd_opensdk_java/src/javatype.rs
//! javatype.ts — TypeRef → Java type/decode expressions, query kind, const &
//! union helpers. Operates on serde_json::Value TypeRefs.

use serde_json::Value;
use std::collections::HashSet;

use crate::ir::{str_field, Types};
use crate::jsrt::{json_str, pascal_case};

/// (property, mapping) of a mapped union, else None.
pub fn union_mapping(ty: &Value) -> Option<(String, Vec<(String, String)>)> {
    if str_field(ty, "kind") != Some("union") {
        return None;
    }
    let disc = ty.get("discriminator")?;
    let prop = str_field(disc, "propertyName")?;
    let mapping = disc.get("mapping")?.as_object()?;
    if mapping.is_empty() {
        return None;
    }
    let entries = mapping
        .iter()
        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
        .collect();
    Some((prop.to_string(), entries))
}

pub fn is_const_field(field: &Value) -> bool {
    let ty = field.get("type");
    ty.and_then(|t| str_field(t, "kind")) == Some("scalar")
        && ty.and_then(|t| t.get("const")).is_some()
}

/// The Java literal for a const scalar value.
pub fn const_literal(value: &Value) -> String {
    match value {
        Value::Number(_) => format!("{value}L"),
        Value::Bool(b) => b.to_string(),
        other => json_str(&js_string(other)),
    }
}

/// JS `String(value)`.
fn js_string(v: &Value) -> String {
    match v {
        Value::Null => "null".to_string(),
        Value::Bool(b) => b.to_string(),
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        _ => v.to_string(),
    }
}

pub fn is_binary_type_ref(ref_: Option<&Value>, types: &Types, seen: &mut HashSet<String>) -> bool {
    let Some(r) = ref_ else { return false };
    match str_field(r, "kind") {
        Some("scalar") => {
            str_field(r, "scalar") == Some("string") && str_field(r, "format") == Some("binary")
        }
        Some("array") => is_binary_type_ref(r.get("items"), types, seen),
        Some("ref") => {
            let Some(name) = str_field(r, "name") else {
                return false;
            };
            if seen.contains(name) {
                return false;
            }
            seen.insert(name.to_string());
            let Some(named) = types.get(name) else {
                return false;
            };
            match str_field(named, "kind") {
                Some("union") => named
                    .get("variants")
                    .and_then(|v| v.as_array())
                    .map(|vs| vs.iter().any(|v| is_binary_type_ref(Some(v), types, seen)))
                    .unwrap_or(false),
                Some("alias") => is_binary_type_ref(named.get("of"), types, seen),
                _ => false,
            }
        }
        _ => false,
    }
}

#[derive(PartialEq)]
pub enum QueryKind {
    Scalar,
    Array,
    Map,
    Object,
}

pub fn query_kind(ref_: Option<&Value>, types: &Types) -> QueryKind {
    let Some(r) = ref_ else {
        return QueryKind::Scalar;
    };
    match str_field(r, "kind") {
        Some("scalar") => QueryKind::Scalar,
        Some("array") => QueryKind::Array,
        Some("map") => QueryKind::Map,
        Some("ref") => {
            let name = str_field(r, "name");
            let named = name.and_then(|n| types.get(n));
            match named.and_then(|n| str_field(n, "kind")) {
                Some("enum") => QueryKind::Scalar,
                Some("alias") => query_kind(named.unwrap().get("of"), types),
                _ => QueryKind::Object,
            }
        }
        _ => QueryKind::Object,
    }
}

pub fn java_type(ref_: Option<&Value>, types: &Types) -> String {
    let Some(r) = ref_ else {
        return "Object".to_string();
    };
    match str_field(r, "kind") {
        Some("scalar") => java_scalar(str_field(r, "scalar"), str_field(r, "format")),
        Some("array") => format!("List<{}>", java_type(r.get("items"), types)),
        Some("map") => format!("Map<String, {}>", java_type(r.get("values"), types)),
        Some("ref") => java_ref_type(str_field(r, "name"), types),
        _ => "Object".to_string(),
    }
}

fn java_scalar(scalar: Option<&str>, format: Option<&str>) -> String {
    match scalar {
        Some("string") => if format == Some("binary") {
            "byte[]"
        } else {
            "String"
        }
        .to_string(),
        Some("integer") => "Long".to_string(),
        Some("number") => "Double".to_string(),
        Some("boolean") => "Boolean".to_string(),
        _ => "Object".to_string(),
    }
}

fn java_ref_type(name: Option<&str>, types: &Types) -> String {
    let Some(name) = name else {
        return "Object".to_string();
    };
    let Some(named) = types.get(name) else {
        return "Object".to_string();
    };
    match str_field(named, "kind") {
        Some("struct") | Some("enum") => pascal_case(name),
        Some("alias") => java_type(named.get("of"), types),
        _ => "Object".to_string(),
    }
}

fn lambda_var(depth: usize) -> String {
    if depth == 0 {
        "__e".to_string()
    } else {
        format!("__e{depth}")
    }
}

/// A Java expression decoding a parsed-JSON `Object` `src` into `ref`'s type.
pub fn java_decode(ref_: Option<&Value>, src: &str, types: &Types, depth: usize) -> String {
    let Some(r) = ref_ else {
        return src.to_string();
    };
    match str_field(r, "kind") {
        Some("scalar") => scalar_decode(str_field(r, "scalar"), str_field(r, "format"), src),
        Some("array") => {
            let v = lambda_var(depth);
            format!(
                "Json.mapList({src}, {v} -> {})",
                java_decode(r.get("items"), &v, types, depth + 1)
            )
        }
        Some("map") => {
            let v = lambda_var(depth);
            format!(
                "Json.mapValues({src}, {v} -> {})",
                java_decode(r.get("values"), &v, types, depth + 1)
            )
        }
        Some("ref") => ref_decode(str_field(r, "name"), src, types),
        _ => src.to_string(),
    }
}

fn scalar_decode(scalar: Option<&str>, format: Option<&str>, src: &str) -> String {
    match scalar {
        Some("string") => {
            if format == Some("binary") {
                format!("Json.asBytes({src})")
            } else {
                format!("Json.asString({src})")
            }
        }
        Some("integer") => format!("Json.asLong({src})"),
        Some("number") => format!("Json.asDouble({src})"),
        Some("boolean") => format!("Json.asBoolean({src})"),
        _ => src.to_string(),
    }
}

fn ref_decode(name: Option<&str>, src: &str, types: &Types) -> String {
    let Some(name) = name else {
        return src.to_string();
    };
    let Some(named) = types.get(name) else {
        return src.to_string();
    };
    match str_field(named, "kind") {
        Some("struct") | Some("enum") => format!("{}.fromJson({src})", pascal_case(name)),
        Some("alias") => java_decode(named.get("of"), src, types, 0),
        Some("union") if union_mapping(named).is_some() => {
            format!("{}.fromJson({src})", pascal_case(name))
        }
        _ => src.to_string(),
    }
}
